attributable_error_decrypt.go

```diff
@@ -0,0 +1,207 @@
+// error onion decrypter.
+func NewOnionAttributableErrorDecrypter(circuit *Circuit,
+	structure *AttributableErrorStructure) *OnionAttributableErrorDecrypter {
+
+	return &OnionAttributableErrorDecrypter{
+		attributableErrorBase: newAttributableErrorBase(structure),
+		circuit:               circuit,
+	}
+}
+
+// DecryptError attempts to decrypt the passed encrypted error response. The
+// onion failure is encrypted in backward manner, starting from the node where
+// error have occurred. As a result, in order to decrypt the error we need get
+// all shared secret and apply decryption in the reverse order. A structure is
+// returned that contains the decrypted error message and information on the
+// sender.
+func (o *OnionAttributableErrorDecrypter) DecryptError(encryptedData []byte) (
+	*DecryptedAttributableError, error) {
+
+	// Ensure the error message length is enough to contain the payloads and
+	// hmacs blocks. Otherwise blame the first hop.
+	if len(encryptedData) < minOnionErrorLength+o.hmacsAndPayloadsLen {
+		return &DecryptedAttributableError{
+			DecryptedError: DecryptedError{
+				SenderIdx: 1,
+				Sender:    o.circuit.PaymentPath[0],
+			},
+		}, nil
+	}
+
+	sharedSecrets, err := generateSharedSecrets(
+		o.circuit.PaymentPath,
+		o.circuit.SessionKey,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("error generating shared secret: %w", err)
+	}
+
+	var (
+		sender      int
+		msg         []byte
+		dummySecret Hash256
+	)
+	copy(dummySecret[:], bytes.Repeat([]byte{1}, 32))
+
+	// We'll iterate a constant amount of hops to ensure that we don't give away
+	// an timing information pertaining to the position in the route that the
+	// error emanated from.
+	hopPayloads := make([][]byte, 0)
+	for i := 0; i < o.maxHops; i++ {
+		var sharedSecret Hash256
+
+		// If we've already found the sender, then we'll use our dummy secret to
+		// continue decryption attempts to fill out the rest of the loop.
+		// Otherwise, we'll use the next shared secret in line.
+		if sender != 0 || i > len(sharedSecrets)-1 {
+			sharedSecret = dummySecret
+		} else {
+			sharedSecret = sharedSecrets[i]
+		}
+
+		// With the shared secret, we'll now strip off a layer of encryption
+		// from the encrypted error payload.
+		encryptedData = onionEncrypt(&sharedSecret, encryptedData)
+
+		message, payloads, hmacs := o.getMsgComponents(encryptedData)
+
+		expectedHmac := o.calculateHmac(
+			sharedSecret, i, message, payloads, hmacs,
+		)
+		actualHmac := hmacs[i*sha256.Size : (i+1)*sha256.Size]
+
+		// If the hmac does not match up, exit with a nil message but only after
+		// finishing the constant number of iterations.
+		if !bytes.Equal(actualHmac, expectedHmac) && sender == 0 {
+			sender = i + 1
+			msg = nil
+		}
+
+		// Extract the payload and exit with a nil message if it is invalid.
+		source, payload, err := o.extractPayload(payloads)
+		if sender == 0 {
+			if err != nil {
```

Comment:
So we don't handle the error at all if the sender is known?

Reply:
The handling consists of returning the current position to the caller.

Some invalid data was encountered because we couldn't extract the payload. This means that there is no actual message to return. With the position information, the caller can apply a selective penalty.